File: android/app/src/main/kotlin/com/baseapp/android/components/feeds/ProjectFeedService.kt
package com.baseapp.android.components.feeds

import com.apollographql.apollo.api.Optional
import com.baseapp.android.components.gql
import com.baseapp.android.graphql.ProjectRowData
import com.baseapp.android.graphql.SortDirectionValue
import com.baseapp.android.graphql.generated.type.ConnectionInput
import com.baseapp.android.graphql.generated.type.PaginationInput
import com.baseapp.android.graphql.generated.type.ProjectConnectionFilters
import com.baseapp.android.graphql.generated.type.ProjectConnectionInput
import com.baseapp.android.graphql.generated.type.SortOrderInput
import com.baseapp.android.query.ConnectionPage

class ProjectFeedFilters {
    var search: String? = null
    var sortFieldName: String = "name"
    var sortDirection: SortDirectionValue = SortDirectionValue.asc
}

/**
 * Thin translation layer between feed filters and the generated connection
 * inputs. Mirrors the web app's Projects query (sort by name, ascending).
 */
class ProjectFeedService {
    suspend fun fetchPage(
        filters: ProjectFeedFilters,
        pageSize: Int,
        afterCursor: String?,
    ): ConnectionPage<ProjectRowData> {
        val searchValue = filters.search?.trim() ?: ""
        val result = gql.fetchProjects(
            ProjectConnectionInput(
                filters = if (searchValue.isEmpty()) {
                    Optional.absent()
                } else {
                    Optional.present(
                        ProjectConnectionFilters(name = Optional.present(searchValue))
                    )
                },
                connection = ConnectionInput(
                    pagination = PaginationInput(
                        first = pageSize,
                        after = afterCursor?.let { Optional.present(it) } ?: Optional.absent(),
                    ),
                    sort = listOf(
                        SortOrderInput(
                            fieldName = filters.sortFieldName,
                            direction = filters.sortDirection,
                        )
                    ),
                ),
            )
        )

        return ConnectionPage(
            items = result.nodes.mapNotNull { it?.projectFields },
            endCursor = result.pageInfo.pageInfoFields.endCursor,
            hasNextPage = result.pageInfo.pageInfoFields.hasNextPage,
        )
    }
}

File: ios/App/Components/Feeds/ProjectFeed/ProjectFeedService.swift
import Foundation
import AppGraphqlApi

final class ProjectFeedFilters {
  var search: String?
  var sortFieldName: String = "name"
  var sortDirection: SortDirection = .asc
}

struct ProjectFeedPage {
  let rows: [ProjectRowData]
  let endCursor: String?
  let hasNextPage: Bool
}

@MainActor
final class ProjectFeedService {
  private let operations = ProjectOperations()

  func fetchPage(
    filters: ProjectFeedFilters,
    pageSize: Int,
    afterCursor: String?
  ) async throws -> ConnectionPage<ProjectRowData> {
    let page = try await operations.fetchProjects(
      filters: filters,
      pageSize: pageSize,
      afterCursor: afterCursor
    )

    return ConnectionPage(
      items: page.rows,
      endCursor: page.endCursor,
      hasNextPage: page.hasNextPage
    )
  }
}
